scripts/test.sh

```diff
@@ -5,13 +5,161 @@ set -ex
+  exit 1
+fi
+
+# Check for changed Rust files if local is specified
+if [ "$mode" = "local" ] && [ -z "$changed_files" ]; then
+  echo "No files provided for local mode. Either no changes were made or the files were not provided."
+  exit 0
+fi
+
+# Prepare apps
 ./build-all-apps.sh
 
 # Prepare contracts
 ../contracts/registry/build.sh
 ../contracts/context-config/build.sh
 ../contracts/proxy-lib/build-test-deps.sh
 
-# Run cargo test
-cargo test
+# Run all tests
+if [ "$mode" = "all" ]; then
+  echo "Running all tests..."
+  cargo +nightly test
+  exit 0
+fi
+
+# Run tests for changed files
+echo "The following Rust files have changed:" "${changed_files[@]}"
+# Find affected crates
+changed_crates=()
+# List of crates defined in workspace
+all_crates=($(awk '/\[workspace\]/,/^\s*$/' ../Cargo.toml | grep -E 'members' -A 100 | grep -Eo '"[^"]*"' | tr -d '"' | sed 's|^\./||'))
+echo "All workspace crates" "${all_crates[@]}"
+
+for file in "${changed_files[@]}"; do
+    # Extract the crate name by stripping the path to get the crate folder
+    crate_name=$(echo "$file" | sed -E 's|^(crates/[^/]+)/.*|\1|')
+    changed_crates+=("$crate_name")
+done
+echo "Detected crates from changed files" "${changed_crates[@]}"
+
+# Remove duplicates
+unique_changed_crates=()
+for item in "${changed_crates[@]}"; do
+    # Check if item is already in unique_array
+    duplicate=false
+    for unique_item in "${unique_changed_crates[@]}"; do
+        if [[ "$item" == "$unique_item" ]]; then
+            duplicate=true
+            break
+        fi
+    done
+    # If item is not a duplicate, add it to unique_array
+    if ! $duplicate; then
+        unique_changed_crates+=("$item")
+    fi
+done
+echo "Unique crates from changed files" "${unique_changed_crates[@]}"
+
+# array of dependencies in format crates/crate
+dep_arr=()
+for changed_crate in "${unique_changed_crates[@]}"; do
+    # Replace "crates/" with "calimero-" to find dependencies as dependencies are imported as calimero-name from crates/name
+    # e.g. In list I have crates/node-primitives but as dependency it is calimero-node-primitives.
+    calimero_package_name+=("${changed_crate/crates\//calimero-}")
+    # get list of all dependencies from the crate including external dependencies
+    dependencies=($(cargo metadata --format-version=1 --no-deps | jq -r --arg CRATE "$calimero_package_name" '
+        .packages[] |
+        select(.dependencies | any(.name == $CRATE)) |
+        .name
+    '))
+
+    for dep in "${dependencies[@]}"; do
+        # Compare dependency with list of crates from workspace to skip external dependencies
+        calimero_dep_crate_path="${dep/calimero-/crates/}"
+        for crate in "${all_crates[@]}"; do
+            if [[ "$crate" == "$calimero_dep_crate_path" ]]; then
+                echo "Found matching dependency: $calimero_dep_crate_path"
+                dep_arr+=("$calimero_dep_crate_path")
+            fi
+        done
+    done
```

Comment:
we shouldn't need to test dependencies of changed crates, but "dependents" of changed crates.

`calimero-node` depends on `calimero-context` which depends on `calimero-primitives`, if `calimero-context` changes we need to test that as well as `calimero-node`, but not necessarily `calimero-primitives`